### ISSUE TITLE
VS 2015 build failed

### ISSUE THREAD

**Issue by roy81f** [takuya-takeuchi/DlibDotNet#20]
When trying to build dlibdotnet on VS 2015 i got many errors (both cmd and vs).
Any suggestions on how I can fix this?
( Win10 / VS2015 / DLIB-19.13 / CUDA9.2 + CUDNN )

-----------------------------------------------------------------
  DlibDotNet.Native.dir\Release\/dlib/vector.cpp.obj
CMakeCXXCompilerId.cpp.obj : error LNK2005: main already defined in CMakeCXXCompilerId.cpp.obj [C:\DlibDotNet\src\DlibD
otNet.Native\build\DlibDotNet.Native.vcxproj]
CMakeCXXCompilerId.cpp.obj : error LNK2005: "char const * const info_compiler" (?info_compiler@@3PEBDEB) already define
d in CMakeCXXCompilerId.cpp.obj [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
CMakeCXXCompilerId.cpp.obj : error LNK2005: "char const * const info_platform" (?info_platform@@3PEBDEB) already define
d in CMakeCXXCompilerId.cpp.obj [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
CMakeCXXCompilerId.cpp.obj : error LNK2005: "char const * const info_arch" (?info_arch@@3PEBDEB) already defined in CMa
keCXXCompilerId.cpp.obj [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
CMakeCXXCompilerId.cpp.obj : error LNK2005: "char const * const info_language_dialect_default" (?info_language_dialect_
default@@3PEBDEB) already defined in CMakeCXXCompilerId.cpp.obj [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.N
ative.vcxproj]
     Creating library C:/DlibDotNet/src/DlibDotNet.Native/build/Release/DlibDotNet.Native.lib and object C:/DlibDotNet/
  src/DlibDotNet.Native/build/Release/DlibDotNet.Native.exp
C:\DlibDotNet\src\DlibDotNet.Native\build\Release\DlibDotNet.Native.dll : fatal error LNK1169: one or more multiply def
ined symbols found [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
Done Building Project "C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj" (default targets) -- FAILED
.

Done Building Project "C:\DlibDotNet\src\DlibDotNet.Native\build\ALL_BUILD.vcxproj" (default targets) -- FAILED.

Build FAILED.

"C:\DlibDotNet\src\DlibDotNet.Native\build\ALL_BUILD.vcxproj" (default target) (1) ->
"C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj" (default target) (3) ->
(ClCompile target) ->
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\ppltaskscheduler.h(52): warning C4355: 'this': used in
 base member initializer list [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\pplwin.h(60): warning C4355: 'this': used in base memb
er initializer list [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\ppltasks.h(1467): warning C4355: 'this': used in base
member initializer list [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\ppltaskscheduler.h(52): warning C4355: 'this': used in
 base member initializer list [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\pplwin.h(60): warning C4355: 'this': used in base memb
er initializer list [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
  C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include\ppltasks.h(1467): warning C4355: 'this': used in base
member initializer list [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]

"C:\DlibDotNet\src\DlibDotNet.Native\build\ALL_BUILD.vcxproj" (default target) (1) ->
"C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj" (default target) (3) ->
(Link target) ->
  CMakeCXXCompilerId.cpp.obj : error LNK2005: main already defined in CMakeCXXCompilerId.cpp.obj [C:\DlibDotNet\src\Dli
bDotNet.Native\build\DlibDotNet.Native.vcxproj]
  CMakeCXXCompilerId.cpp.obj : error LNK2005: "char const * const info_compiler" (?info_compiler@@3PEBDEB) already defi
ned in CMakeCXXCompilerId.cpp.obj [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
  CMakeCXXCompilerId.cpp.obj : error LNK2005: "char const * const info_platform" (?info_platform@@3PEBDEB) already defi
ned in CMakeCXXCompilerId.cpp.obj [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
  CMakeCXXCompilerId.cpp.obj : error LNK2005: "char const * const info_arch" (?info_arch@@3PEBDEB) already defined in C
MakeCXXCompilerId.cpp.obj [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]
  CMakeCXXCompilerId.cpp.obj : error LNK2005: "char const * const info_language_dialect_default" (?info_language_dialec
t_default@@3PEBDEB) already defined in CMakeCXXCompilerId.cpp.obj [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet
.Native.vcxproj]
  C:\DlibDotNet\src\DlibDotNet.Native\build\Release\DlibDotNet.Native.dll : fatal error LNK1169: one or more multiply d
efined symbols found [C:\DlibDotNet\src\DlibDotNet.Native\build\DlibDotNet.Native.vcxproj]

    6 Warning(s)
    6 Error(s)

**Comment by takuya-takeuchi** [comment]
I tried the following command to build by VS2015.

````
mkdir build
cd build
cmake -G "Visual Studio 14 2015 Win64" ^
           -DDLIB_PATH=D:/Works/Lib/DLib/19.13 ^
           ..
cmake --build . --config Release
````

But there is no error.

I think you face to same issue #5.
But I tried the backslash for cmake command param and it can not reproduce same error.
I have no any idea for now.

Does it success if disable cuda?

**Comment by roy81f** [comment]
i follow the tutorial, the first command  (cmake -G "Visual Studio 14 2015 Win64" ^
           -DDLIB_PATH=D:/Works/Lib/DLib/19.13 ^) working without any issue.

i didn't compiled without Cuda yet but i will try.

Thanks for a quick reply.

**Comment by roy81f** [comment]
BTW, in VS2015 i got errors from DlibDotNet.csproj , its seems the project file should compile on 2017 only.

**Comment by takuya-takeuchi** [comment]
Yes, DlibDotNet aims to support cross platform.
So it use .NET Starndard 2.0.
VS2015 does not compile it.

Please refer [Ship .NET 4.6.1 support for .NET Standard 2.0 for VS 2015](https://github.com/dotnet/cli/issues/6739)

**Comment by roy81f** [comment]
ok, i installed Vs2017,   build DLIB successfully,  from some reason i can't build the DlibDotNet
via VS2017 its almost done but i got 1 error:

Severity	Code	Description	Project	File	Line	Suppression State
Error		The command "taskkill /F /IM vstest.executionengine.exe /FI "MEMUSAGE gt 1"
xcopy "C:\DlibDotNet\src\DlibDotNet.Native\build\Debug\DlibDotNet.Native.dll" "C:\DlibDotNet\test\DlibDotNet.Tests\bin\Debug\" /y
if "Debug" == "Debug" (
xcopy "C:\DlibDotNet\src\DlibDotNet.Native\build\Debug\DlibDotNet.Native.pdb" "C:\DlibDotNet\test\DlibDotNet.Tests\bin\Debug\" /y
)
xcopy "C:\DlibDotNet\src\DlibDotNet.Native\build\Debug\DlibDotNet.Native.dll" "C:\DlibDotNet\test\DlibDotNet.Tests\TestDeployItem\Debug" /y
if "Debug" == "Debug" (
xcopy "C:\DlibDotNet\src\DlibDotNet.Native\build\Debug\DlibDotNet.Native.pdb" "C:\DlibDotNet\test\DlibDotNet.Tests\TestDeployItem\Debug" /y
)" exited with code 4.	DlibDotNet.Tests

**Comment by takuya-takeuchi** [comment]
Thank you and I’m sorry to trouble you.
The error you face is no problem. It is post build process error.

In my env, vs unit test process may be unstable and post build kill it.

**Comment by takuya-takeuchi** [comment]
Ok, I re-tried comiling with CUDA.
As you know, VS2017 may not support CUDA for now.
Perhaps, this issue only for newer VS2017? It is not for sure...
* FYI
https://devtalk.nvidia.com/default/topic/1027299/cuda-setup-and-installation/cuda-9-failed-to-support-the-latest-visual-studio-2017-version-15-5/

However, it was successful to build with CUDA by VS2015.
Could you try the following command after delete build dir?

````
cmake -G "Visual Studio 14 2015 Win64" ^
           -DDLIB_PATH=D:/Works/Lib/DLib/19.13 ^
           -DCMAKE_PREFIX_PATH=<cuDNN_DIR> ^
           ..
````

***<cuDNN_DIR>*** is cuDNN directory and contains the following objects.
* bin
* include
* lib
* NVIDIA_SLA_cuDNN_Support.txt

But binary size is same DlibDotNet.Native.dll without cuda.
It may be current source code does not functions which uses cuda.

In conclusion, you can use VS2015 for DlibDotNet.Native and use VS2017 for DlibDotNet.

**Comment by roy81f** [comment]
ok, i did better than that,    Install clean Win10 64,  Vs1017, Cuda, build Dlib (without Cuda), than i build the dlibdotnet-master.  could you advice me whats wrong? or better,  could you help me via teamviewer?   i would like to donate when it works.

*** Its seems some dll were created, i can use DlibDotNet.dll, DlibDotNet.Extensions but DlibDotNet.Native.dll is invalid (can't add the file as reference)  ******


in VS2017 i got post build error:

The command "taskkill /F /IM vstest.executionengine.exe /FI "MEMUSAGE gt 1"
xcopy "C:\DlibDotNet17\src\DlibDotNet.Native\build\Debug\DlibDotNet.Native.dll" "C:\DlibDotNet17\test\DlibDotNet.Tests\bin\Debug\" /y
if "Debug" == "Debug" (
xcopy "C:\DlibDotNet17\src\DlibDotNet.Native\build\Debug\DlibDotNet.Native.pdb" "C:\DlibDotNet17\test\DlibDotNet.Tests\bin\Debug\" /y
)
xcopy "C:\DlibDotNet17\src\DlibDotNet.Native\build\Debug\DlibDotNet.Native.dll" "C:\DlibDotNet17\test\DlibDotNet.Tests\TestDeployItem\Debug" /y
if "Debug" == "Debug" (
xcopy "C:\DlibDotNet17\src\DlibDotNet.Native\build\Debug\DlibDotNet.Native.pdb" "C:\DlibDotNet17\test\DlibDotNet.Tests\TestDeployItem\Debug" /y
)" exited with code 2.	DlibDotNet.Tests

**Comment by roy81f** [comment]
i will try,  many thanks.

**Comment by takuya-takeuchi** [comment]
Oh, not at all.
I appreciate your continued support!

**Comment by roy81f** [comment]
im so sorry for all troubles but i hope we are close to solution :-)
i ignored the tests and all compiled,  when i'm trying to run FaceLandmarkDetection example i got the error below
in line:
this._ShapePredictor = new ShapePredictor(@"c:\shape_predictor_68_face_landmarks.dat");

System.DllNotFoundException: 'Unable to load DLL 'DlibDotNet.Native.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)'

from some reason as happened before,  DlibDotNet.dll and DlibDotNet.Extension.dll are compiled but DlibDotNet.Native.dll didn't.

**Comment by takuya-takeuchi** [comment]
> from some reason as happened before, DlibDotNet.dll and DlibDotNet.Extension.dll are compiled but DlibDotNet.Native.dll didn't.

Does it mean that DlibDotNet.Native.dll is not in DlibDotNet\examples\FaceLandmarkDetection\bin\Debug(or Release)\netcoreapp2.0?
DlibDotNet.Native.dll will NOT copy automatically to under the output of examples project directory.

**Comment by roy81f** [comment]
no,  the file is not exist there...
i will try to do everything from the beginning again.

Thank you!

**Comment by takuya-takeuchi** [comment]
There is no additional report.
